src/audio/copier.c

```diff
@@ -0,0 +1,666 @@
+				ipc_pipe->pipeline->source_comp = dev;
+			else
+				ipc_pipe->pipeline->sink_comp = dev;
+
+			break;
+		case ipc4_hda_link_output_class:
+		case ipc4_hda_link_input_class:
+		case ipc4_dmic_link_input_class:
+		case ipc4_i2s_link_output_class:
+		case ipc4_i2s_link_input_class:
+		case ipc4_alh_link_output_class:
+		case ipc4_alh_link_input_class:
+		cd->dai = create_dai(config, copier, &node_id, ipc_pipe->pipeline);
+		if (!cd->dai) {
+			comp_cl_err(&comp_copier, "unenable to create dai");
+			goto error_cd;
+		}
+
+		if (cd->direction == SOF_IPC_STREAM_PLAYBACK)
+			ipc_pipe->pipeline->sink_comp = dev;
+		else
+			ipc_pipe->pipeline->source_comp = dev;
+
+		break;
+		default:
+			comp_cl_err(&comp_copier, "unsupported dma type %x",
+				    (uint32_t)node_id.f.dma_type);
+			goto error_cd;
+		};
+	}
+
+	dev->direction = cd->direction;
+	dev->state = COMP_STATE_READY;
+	return dev;
+
+error_cd:
+	rfree(cd);
+error:
+	rfree(dev);
+	return NULL;
+}
+
+static void copier_free(struct comp_dev *dev)
+{
+	struct copier_data *cd = comp_get_drvdata(dev);
+
+	if (cd->host)
+		cd->host->drv->ops.free(cd->host);
+	else if (cd->dai)
+		cd->dai->drv->ops.free(cd->dai);
+
+	rfree(cd);
+	rfree(dev);
+}
+
+static bool use_no_container_convert_function(enum sof_ipc_frame in,
+					      enum sof_ipc_frame valid_in_bits,
+					      enum sof_ipc_frame out,
+					      enum sof_ipc_frame valid_out_bits)
+{
+	if (in == valid_in_bits && out == valid_out_bits)
+		return true;
+
+	if (valid_in_bits == SOF_IPC_FRAME_S24_4LE && in == SOF_IPC_FRAME_S32_LE &&
+	    valid_out_bits == SOF_IPC_FRAME_S32_LE && out == SOF_IPC_FRAME_S32_LE)
+		return true;
+
+	if (valid_in_bits == SOF_IPC_FRAME_S32_LE && in == SOF_IPC_FRAME_S32_LE &&
+	    valid_out_bits == SOF_IPC_FRAME_S24_4LE && out == SOF_IPC_FRAME_S32_LE)
+		return true;
+
+	return false;
+}
+
+static pcm_converter_func get_converter_func(struct ipc4_audio_format *in_fmt,
+					     struct ipc4_audio_format *out_fmt)
+{
+	enum sof_ipc_frame in, in_valid, out, out_valid;
+
+	in = convert_fmt(in_fmt->depth);
+	in_valid = convert_fmt(in_fmt->valid_bit_depth);
+	out = convert_fmt(out_fmt->depth);
+	out_valid = convert_fmt(out_fmt->valid_bit_depth);
+
+	if (use_no_container_convert_function(in, in_valid, out, out_valid))
+		return pcm_get_conversion_function(in, out);
+	else
+		return pcm_get_conversion_vc_function(in, in_valid, out, out_valid);
+}
+
+static int copier_prepare(struct comp_dev *dev)
+{
+	struct copier_data *cd = comp_get_drvdata(dev);
+	struct comp_buffer *buffer;
+	int ret, dir;
+
+	comp_info(dev, "copier_prepare()");
+
+	/* cannot configure DAI while active */
+	if (dev->state == COMP_STATE_ACTIVE) {
+		comp_info(dev, "copier_config_prepare(): Component is in active state.");
+		return 0;
+	}
+
+	ret = comp_set_state(dev, COMP_TRIGGER_PREPARE);
+	if (ret < 0)
+		return ret;
+
+	if (ret == COMP_STATUS_STATE_ALREADY_SET)
+		return PPL_STATUS_PATH_STOP;
+
+	if (!list_is_empty(&dev->bsink_list)) {
```

Comment:
Not following this logic (and I can see this block in several places), can we have comments describing what we are doing.